glue/core/subset.py

```diff
@@ -1091,6 +1091,81 @@ def __str__(self):
     def __repr__(self):
         return '<%s: %s>' % (self.__class__.__name__, self)
 
+class RoiSubsetState3d(SubsetState):
+    """Subset state for a roi that implements .contains3d
+    """
+
+    @contract(xatt='isinstance(ComponentID)', yatt='isinstance(ComponentID)', zatt='isinstance(ComponentID)')
+    def __init__(self, xatt=None, yatt=None, zatt=None, roi=None):
+        super(RoiSubsetState3d, self).__init__()
+        self.xatt = xatt
+        self.yatt = yatt
+        self.zatt = zatt
+        self.roi = roi
+
+    @property
+    def attributes(self):
+        return (self.xatt, self.yatt, self.zatt)
+
+    @contract(data='isinstance(Data)', view='array_view')
+    def to_mask(self, data, view=None):
+
+        # TODO: make sure that pixel components don't actually take up much
+        #       memory and are just views
+        x = data[self.xatt, view]
+        y = data[self.yatt, view]
+        z = data[self.zatt, view]
+
+        # if (x.ndim == data.ndim and
```

Comment:
I think you can basically ignore/remove this - it was an optimization that was useful when selecting a 2D region in an n-d cube in the image viewer, but since the 3D viewers don't (yet) support showing n-dimensional datasets, this isn't useful. The only thing you need to properly support cubes is to make ``contains3d`` work with n-dimensional arrays.